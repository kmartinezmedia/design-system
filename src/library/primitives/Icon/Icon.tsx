import React from 'react';
import { Text, StyleSheet } from 'react-native';

import { useSpacing, useForeground } from '@designSystem/hooks';
import { SurfaceColorMap, Spacing } from '@designSystem/types';

import {
  glyphMapLarge,
  glyphMapMedium,
  glyphMapSmall,
} from '@designSystem/assets/icons';

const iconMap = {
  small: glyphMapSmall,
  medium: glyphMapMedium,
  large: glyphMapLarge,
} as const;

type IconSizeNameMap = typeof iconMap;

export const Icon = <
  Size extends keyof IconSizeNameMap,
  Name extends keyof IconSizeNameMap[Size],
  Surface extends keyof SurfaceColorMap,
  Color extends SurfaceColorMap[Surface]
>({
  size,
  name,
  spacing = undefined,
  surface = 'background' as Surface,
  color = 'primary' as Color,
}: {
  size: Size;
  name: Name;
  spacing?: Spacing;
  surface?: Surface;
  color?: Color;
}) => {
  const space = useSpacing(spacing);
  const iconColor = useForeground(surface, color);

  return (
    <Text style={[space, styles[size], { color: iconColor }]}>
      {iconMap[size][name]}
    </Text>
  );
};

const styles = StyleSheet.create({
  small: {
    fontFamily: 'CoinbaseIconsSmall',
    fontSize: 12,
  },
  medium: {
    fontFamily: 'CoinbaseIconsMedium',
    fontSize: 16,
  },
  large: {
    fontFamily: 'CoinbaseIconsLarge',
    fontSize: 32,
  },
});
