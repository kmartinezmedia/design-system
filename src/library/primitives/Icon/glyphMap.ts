export const glyphMap = {
  account: '',
  accountPreferences: '',
  addBank: '',
  administrator: '',
  alert: '',
  allTimeHigh: '',
  api: '',
  auditor: '',
  authorizedSignatory: '',
  bank: '',
  barChart: '',
  blockChainAlt: '',
  blockchain: '',
  buy: '',
  buyBitcoin: '',
  camera: '',
  cancel: '',
  card3ds: '',
  checkmark: '',
  checkmarkFilled: '',
  chevron: '',
  circulatingSupply: '',
  clock: '',
  close: '',
  code: '',
  coin: '',
  coinAlt: '',
  consensus: '',
  convert: '',
  convert1: '',
  copy: '',
  copyDocument: '',
  creditCard: '',
  creditCardBack: '',
  creditCardFront: '',
  delete: '',
  deposit: '',
  depositCircle: '',
  document: '',
  dollarSign: '',
  edit: '',
  email: '',
  emailConfirmed: '',
  envelope: '',
  externalLink: '',
  fiat: '',
  fiatUsd: '',
  fileUpload: '',
  follow: '',
  following: '',
  gauge: '',
  governance: '',
  hidePassword: '',
  history: '',
  information: '',
  internalTransfer: '',
  international: '',
  license: '',
  licenseBack: '',
  link: '',
  list: '',
  listSquare: '',
  lock: '',
  marketCap: '',
  mobileCamera: '',
  more: '',
  myNumberCard: '',
  networkFees: '',
  paypal: '',
  permissions: '',
  play: '',
  pricesDown: '',
  pricesNeutral: '',
  pricesUp: '',
  qrCode: '',
  questionMark: '',
  receive: '',
  received: '',
  recover: '',
  recurring: '',
  residenceCard: '',
  reviewer: '',
  safe: '',
  scanQrCode: '',
  scheduledTransfers: '',
  search: '',
  security: '',
  sell: '',
  send: '',
  send1: '',
  sent: '',
  settings: '',
  settingsKnob: '',
  share: '',
  shield: '',
  showPassword: '',
  sparkChart: '',
  stake: '',
  stakingReward: '',
  stateId: '',
  tradeInstantly: '',
  transfer: '',
  unlock: '',
  wallet: '',
  walletCircle: '',
  warning: '',
  webcam: '',
  wireTransfer: '',
  withdraw: '',
  withdrawCircle: '',
};
