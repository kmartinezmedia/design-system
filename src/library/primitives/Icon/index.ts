export * from './Icon';
export * from './example';
