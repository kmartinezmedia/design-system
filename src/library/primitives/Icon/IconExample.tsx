import React from 'react';
import { Title1, Label } from '@designSystem/primitives/typography';
import { VStack } from '@designSystem/primitives/spacing';

import {
  glyphMapLarge,
  glyphMapMedium,
  glyphMapSmall,
} from '@designSystem/assets/icons';
import { Icon } from './Icon';

export const IconExample = () => {
  return (
    <VStack spacing={0}>
      <Title1>Icons</Title1>
      {Object.keys(glyphMapLarge).map((name) => (
        <VStack key={name} spacing={{ vertical: 3 }}>
          <Label>{name}</Label>
          <Icon size="large" name={name as keyof typeof glyphMapLarge} />
        </VStack>
      ))}
      {Object.keys(glyphMapMedium).map((name) => (
        <VStack key={name} spacing={{ vertical: 3 }}>
          <Label>{name}</Label>
          <Icon size="medium" name={name as keyof typeof glyphMapMedium} />
        </VStack>
      ))}
      {Object.keys(glyphMapSmall).map((name) => (
        <VStack key={name} spacing={{ vertical: 3 }}>
          <Label>{name}</Label>
          <Icon size="small" name={name as keyof typeof glyphMapSmall} />
        </VStack>
      ))}
    </VStack>
  );
};
