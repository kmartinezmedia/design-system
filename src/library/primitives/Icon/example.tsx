import React from 'react';
import { Title1, Label } from '@designSystem/primitives/typography';
import { VStack } from '@designSystem/primitives/spacing';

import { glyphMap } from './glyphMap';
import { Icon } from './Icon';

export const IconExample = () => {
  return (
    <VStack spacing={0}>
      <Title1>Icons</Title1>
      {Object.keys(glyphMap).map((name) => (
        <VStack key={name} spacing={{ vertical: 3 }}>
          <Label>{name}</Label>
          <Icon name={name as keyof typeof glyphMap} />
        </VStack>
      ))}
    </VStack>
  );
};
