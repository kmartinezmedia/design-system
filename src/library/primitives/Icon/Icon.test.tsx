import React from 'react';
import { create, act, ReactTestRenderer } from 'react-test-renderer';
import { Icon } from './Icon';

describe('Icon', () => {
  it('Large', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(<Icon name="phone" size="large" />);
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('Medium', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(<Icon name="close" size="medium" />);
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });

  it('Small', () => {
    let component: ReactTestRenderer;
    act(() => {
      component = create(<Icon name="caretRight" size="small" />);
    });
    expect(component!.toJSON()).toMatchSnapshot();
  });
});
