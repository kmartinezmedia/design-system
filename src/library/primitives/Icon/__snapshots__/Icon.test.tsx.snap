// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon Large 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "fontFamily": "CoinbaseIconsLarge",
        "fontSize": 32,
      },
      Object {
        "color": "#1652F0",
      },
    ]
  }
>
  
</Text>
`;

exports[`Icon Medium 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "fontFamily": "CoinbaseIconsMedium",
        "fontSize": 16,
      },
      Object {
        "color": "#1652F0",
      },
    ]
  }
>
  
</Text>
`;

exports[`Icon Small 1`] = `
<Text
  style={
    Array [
      undefined,
      Object {
        "fontFamily": "CoinbaseIconsSmall",
        "fontSize": 12,
      },
      Object {
        "color": "#1652F0",
      },
    ]
  }
>
  
</Text>
`;
